pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -38,107 +44,227 @@ type GroupTokenBucket struct {
+}
+
 // GroupTokenBucketState is the running state of TokenBucket.
 type GroupTokenBucketState struct {
-	Tokens      float64    `json:"tokens,omitempty"`
+	Tokens float64 `json:"tokens,omitempty"`
+	// ClientUniqueID -> TokenSlot
+	tokenSlots                 map[uint64]*TokenSlot
+	clientConsumptionTokensSum float64
+
 	LastUpdate  *time.Time `json:"last_update,omitempty"`
 	Initialized bool       `json:"initialized"`
 	// settingChanged is used to avoid that the number of tokens returned is jitter because of changing fill rate.
 	settingChanged bool
 }
 
 // Clone returns the copy of GroupTokenBucketState
-func (s *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+func (gts *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+	tokenSlots := make(map[uint64]*TokenSlot)
+	for id, tokens := range gts.tokenSlots {
+		tokenSlots[id] = tokens
+	}
+	var lastUpdate *time.Time
+	if gts.LastUpdate != nil {
+		newLastUpdate := *gts.LastUpdate
+		lastUpdate = &newLastUpdate
+	}
 	return &GroupTokenBucketState{
-		Tokens:      s.Tokens,
-		LastUpdate:  s.LastUpdate,
-		Initialized: s.Initialized,
+		Tokens:                     gts.Tokens,
+		LastUpdate:                 lastUpdate,
+		Initialized:                gts.Initialized,
+		tokenSlots:                 tokenSlots,
+		clientConsumptionTokensSum: gts.clientConsumptionTokensSum,
+	}
+}
+
+func (gts *GroupTokenBucketState) cleanupAssignTokenSum() {
+	gts.clientConsumptionTokensSum = 0
+	for _, slot := range gts.tokenSlots {
+		slot.requireTokensSum = 0
+	}
+}
+
+func (gts *GroupTokenBucketState) balanceSlotTokens(
+	clientUniqueID uint64,
+	settings *rmpb.TokenLimitSettings,
+	consumptionToken float64) {
+	now := time.Now()
+	slot, exist := gts.tokenSlots[clientUniqueID]
+	if !exist {
+		gts.cleanupAssignTokenSum()
+		slot = &TokenSlot{}
+		gts.tokenSlots[clientUniqueID] = slot
+	} else if gts.clientConsumptionTokensSum >= maxAssignTokens {
+		gts.cleanupAssignTokenSum()
+	}
+	slot.lastUpdate = &now
+	slot.requireTokensSum += consumptionToken
+	gts.clientConsumptionTokensSum += consumptionToken
+
+	evenRatio := 1 / float64(len(gts.tokenSlots))
+	if settings.GetBurstLimit() < 0 {
+		for _, slot := range gts.tokenSlots {
+			slot.settings = &rmpb.TokenLimitSettings{
+				FillRate:   uint64(float64(settings.GetFillRate()) * evenRatio),
+				BurstLimit: settings.GetBurstLimit(),
+			}
+		}
+		return
+	}
+
+retryLoop:
+	for clientID, slot := range gts.tokenSlots {
+		// Clean up those slots that have not been used for a long time.
+		if clientID != clientUniqueID && now.Sub(*slot.lastUpdate) >= defaultSlotStalePeriod {
+			delete(gts.tokenSlots, clientID)
+			evenRatio = 1 / float64(len(gts.tokenSlots))
+			gts.cleanupAssignTokenSum()
+			goto retryLoop
+		}
+
+		var ratio float64
+		if gts.clientConsumptionTokensSum == 0 || len(gts.tokenSlots) == 1 {
+			ratio = evenRatio
+		} else {
+			ratio = (1 - slot.requireTokensSum/gts.clientConsumptionTokensSum + evenRatio) * evenRatio
```

Comment:
need add more comments about this formula.

Reply:
done.